### ISSUE TITLE
No JSON response for caught exceptions in `Router::process_http_request()`

### ISSUE THREAD

**Issue by dlh01** [wp-graphql/wp-graphql#442]
Using [the `book` example from the Wiki](https://github.com/wp-graphql/wp-graphql/wiki/Extending-and-Customizing#post-type), but with the `name` arguments omitted:

```php
<?php
add_action( 'init', function() {
	register_post_type( 'book', [
		'label' => __( 'Books', 'wp-graphql-publishers' ),
		'supports' => [ 'title', 'editor', 'custom-fields' ],
		'public' => true,
		'show_in_graphql' => true,
	] );
} );
```

an Exception is thrown, and `Router::process_http_request()` catches it, but doesn't seem to return JSON after that. In my test environment, the home page HTML is returned instead with a `200` response.

I'm using the `develop` branch checked out to 6e349b7beb705f3a5d1a621a2d9089fcf5e22845, and running WordPress trunk with the Twenty Seventeen theme and no other plugins.

**Comment by jasonbahl** [comment]
@dlh01 did you also include `graphql_single_name` and `graphql_plural_name`? I don’t see it in the code snippet

**Comment by dlh01** [comment]
No. I skipped them because I thought that would be an easy way to generate the exception, but maybe I'm misunderstanding what you're getting at.

**Comment by jasonbahl** [comment]
@dlh01 you need to define a single and plural name for the Type, via: 

```
'graphql_single_name' => 'book',
'graphql_plural_name' => 'books',
```

Do you have those defined in your code?

**Comment by dlh01** [comment]
Sorry, I think I wrote my report in a confusing way.

My sample code has a bug on purpose. As I understand it, my sample code should generate an Exception, and a 500 response should be returned by `Router::process_http_request()` with an error message:

https://github.com/wp-graphql/wp-graphql/blob/6e349b7beb705f3a5d1a621a2d9089fcf5e22845/src/Router.php#L463-L473

But from what I can tell, that isn't happening. I get a 200 response and HTML instead:

<img width="1311" alt="screen shot 2018-05-07 at 12 30 58 pm" src="https://user-images.githubusercontent.com/697432/39713034-3648def0-51f3-11e8-82d4-e56feaac1545.png">

So I'm wondering whether there's a missing call to `wp_send_json()`, or something along those lines.

**Comment by jasonbahl** [comment]
@dlh01 ah, I see what you mean. 

In this case, the exception happens outside of the HTTP request, so it's not included in the scope of the try/catch statement in the router that catches exceptions. 

Perhaps we should have the `get_allowed_post_types()` and `get_allowed_taxonomies()` methods fail more gracefully though. 

Maybe instead of throwing an exception, if all 3 fields (show_in_graphql, graphql_single_name, and graphql_plural_name) are not present we simply don't add the post_type/taxonomy to the list of allowed ones, and silently fail if one or two are filled in but not all 3. . .then if GRAPHQL_DEBUG is true, we can `wp_die()` with a message of why the post_type isn't configured properly for GraphQL?

Thoughts?

**Comment by dlh01** [comment]
I'm not sure I have enough experience with the plugin to comment usefully about a change like the one you describe, so I'd be interested in what others have to say about it.

That said, the `register_post_type()` code is just a way to demonstrate the behavior that I'm wondering about. Any kind of Exception that occurs in the try/catch statement seems to cause it. For example:

```diff
diff --git a/src/Router.php b/src/Router.php
index 7f7c990..e39d672 100755
--- a/src/Router.php
+++ b/src/Router.php
@@ -327,6 +327,7 @@ class Router {
 		$user            = null;
 
 		try {
+			throw new \Exception( 'Testing' );
 
 			/**
 			 * Store the global post so it can be reset after GraphQL execution
```

This exception is caught, but the response is still 200 with HTML instead of JSON. There definitely might be a reason for it that I'm overlooking. The fact that `$response['errors']` is set but never used, as far as I can tell, does seem suspect, though: https://github.com/wp-graphql/wp-graphql/blob/6e349b7beb705f3a5d1a621a2d9089fcf5e22845/src/Router.php#L471-L478

**Comment by jasonbahl** [comment]
@dlh01 thanks for the extra details. I think I see what’s up and have an idea on how to fix it.

**Comment by jasonbahl** [comment]
closed by #443
